File: packages/core/src/tool/builtin/python-scripts/security/security.py
#!/usr/bin/env python3
"""授权范围内的网络安全侦察 / 合规审计工具（Python 版）。

    仅限对「自有资产 / 已签署授权书的目标」使用。本脚本不做任何攻击性动作
    （不含 DoS、未授权爆破、RCE、恶意载荷），仅提供被动侦察与合规审计能力：
      - port_scan    TCP 连接端口扫描（开放端口 + 服务推断）
      - http_headers HTTP 安全响应头审计（HSTS/CSP/X-Frame-Options…）
      - ssl_check    TLS 证书审计（颁发者/有效期/剩余天数/协议）
      - dns_lookup   DNS 记录查询（A/AAAA/CNAME/MX/NS/TXT）
      - subdomain    基于字典的子域发现（仅 DNS 解析，不做请求）

    依赖（打包期预烤，离线可用）：requests、dnspython；socket/ssl 为标准库。
    输出：纯文本摘要，直接作为工具结果返回。

    法律声明：使用者须确保其对该目标拥有合法授权。滥用本工具造成的一切后果
    由使用者自行承担。
"""
from __future__ import annotations

import argparse
import json
import socket
import ssl
import sys
from datetime import datetime, timezone

try:
    import requests
except ImportError:
    requests = None

try:
    import dns.resolver
    import dns.exception
except ImportError:
    dns = None


COMMON_PORTS = {
    21: "ftp", 22: "ssh", 23: "telnet", 25: "smtp", 53: "dns", 80: "http",
    110: "pop3", 111: "rpcbind", 135: "msrpc", 139: "netbios", 143: "imap",
    443: "https", 445: "smb", 993: "imaps", 995: "pop3s", 1433: "mssql",
    1521: "oracle", 1723: "pptp", 3306: "mysql", 3389: "rdp", 5432: "postgres",
    5900: "vnc", 6379: "redis", 8080: "http-alt", 8443: "https-alt",
    9200: "elasticsearch", 27017: "mongodb",
}

SECURITY_HEADERS = {
    "Strict-Transport-Security": "HSTS（强制 HTTPS）",
    "Content-Security-Policy": "CSP（防 XSS/注入）",
    "X-Frame-Options": "点击劫持防护",
    "X-Content-Type-Options": "MIME 嗅探防护",
    "Referrer-Policy": "引用来源泄露防护",
    "Permissions-Policy": "浏览器特性权限控制",
    "X-XSS-Protection": "旧版 XSS 过滤器（已弃用但仍常见）",
}

# 常见子域字典（保守规模，避免成为爆破工具）
SUBDOMAIN_WORDLIST = [
    "www", "mail", "webmail", "ftp", "api", "dev", "test", "staging", "beta",
    "admin", "portal", "vpn", "remote", "git", "svn", "ns1", "ns2", "mx",
    "cdn", "static", "assets", "blog", "shop", "app", "m", "mobile", "docs",
    "wiki", "jira", "confluence", "grafana", "jenkins", "ci", "auth", "sso",
]


def _parse_ports(spec: str):
    spec = (spec or "common").strip().lower()
    if spec == "common":
        return sorted(COMMON_PORTS.keys())
    out = set()
    for part in spec.split(","):
        part = part.strip()
        if not part:
            continue
        m = __import__("re").match(r"^(\d+)-(\d+)$", part)
        if m:
            lo = max(1, int(m.group(1)))
            hi = min(65535, int(m.group(2)))
            out.update(range(lo, hi + 1))
        else:
            n = int(part)
            if 1 <= n <= 65535:
                out.add(n)
    return sorted(out)


def cmd_port_scan(a):
    host = a.host
    ports = _parse_ports(a.ports)
    timeout = max(min(int(a.timeout or 1500), 10000), 100)
    conc = max(min(int(a.concurrency or 200), 500), 1)
    print("port_scan %s  ports=%d  timeout=%dms  concurrency=%d" % (host, len(ports), timeout, conc))
    open_ports = []

    import threading
    lock = threading.Lock()

    def scan(p):
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.settimeout(timeout / 1000.0)
        try:
            s.connect((host, p))
            with lock:
                open_ports.append(p)
        except Exception:
            pass
        finally:
            s.close()

    idx = 0
    threads = []
    try:
        while idx < len(ports):
            while len(threads) < conc and idx < len(ports):
                t = threading.Thread(target=scan, args=(ports[idx],))
                t.daemon = True
                t.start()
                threads.append(t)
                idx += 1
            for t in threads:
                t.join(0.05)
            threads = [t for t in threads if t.is_alive()]
    except KeyboardInterrupt:
        pass

    open_ports.sort()
    print("OPEN: %d" % len(open_ports))
    for p in open_ports:
        print("  %d/tcp  %s" % (p, COMMON_PORTS.get(p, "unknown")))


def cmd_http_headers(a):
    if requests is None:
        print("Error: 缺少 requests 依赖（打包环境应预烤）")
        return 1
    url = a.url
    if not url.startswith("http://") and not url.startswith("https://"):
        url = "https://" + url
    try:
        r = requests.get(url, timeout=float(a.timeout or 10), allow_redirects=True,
                         headers={"User-Agent": "Mozilla/5.0 (compatible; YZSecurityAudit/1.0)"})
    except Exception as e:
        print("Error: 请求失败: %s" % e)
        return 1
    print("URL: %s  status=%d  final=%s" % (url, r.status_code, r.url))
    print("响应头安全审计：")
    present = {}
    for k, v in r.headers.items():
        kl = k.lower()
        if kl in (x.lower() for x in SECURITY_HEADERS):
            present[kl] = v
    for hdr, desc in SECURITY_HEADERS.items():
        got = present.get(hdr.lower())
        if got:
            print("  [OK]    %-28s %s" % (hdr, desc))
        else:
            print("  [MISS]  %-28s %s" % (hdr, desc))
    # 服务器/技术指纹（被动，仅读响应头）
    server = r.headers.get("Server") or r.headers.get("X-Powered-By") or "(未暴露)"
    print("SERVER/STACK: %s" % server)
    return 0


def cmd_ssl_check(a):
    host = a.host
    port = int(a.port or 443)
    try:
        ctx = ssl.create_default_context()
        with socket.create_connection((host, port), timeout=float(a.timeout or 10)) as sock:
            with ctx.wrap_socket(sock, server_hostname=host) as ssock:
                cert = ssock.getpeercert()
                proto = ssock.version()
        print("HOST: %s:%d  protocol=%s" % (host, port, proto))
        subj = dict(x[0] for x in cert.get("subject", []))
        issuer = dict(x[0] for x in cert.get("issuer", []))
        print("SUBJECT CN: %s" % subj.get("commonName", "(无)"))
        print("ISSUER:    %s" % issuer.get("organizationName", issuer.get("commonName", "(无)")))
        nb = cert.get("notBefore")
        na = cert.get("notAfter")
        fmt = "%Y%m%d%H%M%SZ"
        try:
            d_na = datetime.strptime(na, fmt).replace(tzinfo=timezone.utc)
            d_nb = datetime.strptime(nb, fmt).replace(tzinfo=timezone.utc)
            now = datetime.now(timezone.utc)
            days_left = (d_na - now).days
            print("VALID FROM: %s" % d_nb.date())
            print("VALID TO:   %s" % d_na.date())
            print("REMAINING:  %d 天 %s" % (days_left, "（即将过期！）" if days_left < 30 else ""))
        except Exception as e:
            print("有效期解析失败: %s" % e)
        sans = cert.get("subjectAltName", [])
        if sans:
            print("SAN: %s" % ", ".join(v for _, v in sans))
    except Exception as e:
        print("Error: SSL 审计失败: %s" % e)
        return 1
    return 0


def cmd_dns_lookup(a):
    if dns is None:
        print("Error: 缺少 dnspython 依赖（打包环境应预烤）")
        return 1
    name = a.name
    rdtype = (a.type or "A").upper()
    try:
        answers = dns.resolver.resolve(name, rdtype)
        print("DNS %s %s:" % (name, rdtype))
        for r in answers:
            print("  %s" % r.to_text())
    except dns.resolver.NXDOMAIN:
        print("Error: 域名不存在 (NXDOMAIN): %s" % name)
        return 1
    except dns.exception.DNSException as e:
        print("Error: DNS 查询失败: %s" % e)
        return 1
    return 0


def cmd_subdomain(a):
    if dns is None:
        print("Error: 缺少 dnspython 依赖（打包环境应预烤）")
        return 1
    domain = a.domain
    found = []
    for sub in SUBDOMAIN_WORDLIST:
        fqdn = "%s.%s" % (sub, domain)
        try:
            dns.resolver.resolve(fqdn, "A")
            found.append(fqdn)
            print("  [+] %s" % fqdn)
        except Exception:
            pass
    print("子域发现完成：命中 %d / 候选 %d（仅 DNS 解析，未发起请求）" % (len(found), len(SUBDOMAIN_WORDLIST)))
    return 0


def main(argv=None):
    p = argparse.ArgumentParser(prog="security", description="授权网络安全侦察/审计（Python）")
    sub = p.add_subparsers(dest="cmd", required=True)

    s = sub.add_parser("port_scan", help="TCP 端口扫描")
    s.add_argument("--host", required=True)
    s.add_argument("--ports", default="common")
    s.add_argument("--timeout", default=1500)
    s.add_argument("--concurrency", default=200)
    s.set_defaults(func=cmd_port_scan)

    s = sub.add_parser("http_headers", help="HTTP 安全头审计")
    s.add_argument("--url", required=True)
    s.add_argument("--timeout", default=10)
    s.set_defaults(func=cmd_http_headers)

    s = sub.add_parser("ssl_check", help="TLS 证书审计")
    s.add_argument("--host", required=True)
    s.add_argument("--port", default=443)
    s.add_argument("--timeout", default=10)
    s.set_defaults(func=cmd_ssl_check)

    s = sub.add_parser("dns_lookup", help="DNS 记录查询")
    s.add_argument("--name", required=True)
    s.add_argument("--type", default="A")
    s.set_defaults(func=cmd_dns_lookup)

    s = sub.add_parser("subdomain", help="子域发现（字典 DNS）")
    s.add_argument("--domain", required=True)
    s.set_defaults(func=cmd_subdomain)

    a = p.parse_args(argv)
    try:
        rc = a.func(a)
    except KeyboardInterrupt:
        print("已中断")
        return 1
    return rc or 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/core/src/tool/builtin/security.ts
// security 内置工具 —— 授权范围内的网络安全侦察 / 合规审计（Python 版）。
// 底层调用随包分发的 python-scripts/security/security.py（requests + dnspython + 标准库）。
// 解释器走 python-runtime 桥：prod 用打包 python-build-standalone，dev/缺失回退系统 python。
// 仅提供被动侦察与合规审计（端口扫描 / HTTP 安全头 / TLS 证书 / DNS / 子域发现），
// 不含任何攻击性动作（DoS / 未授权爆破 / RCE / 恶意载荷）。使用者须对目标拥有合法授权。
import type { BuiltInTool } from '../types';
import type { McpCallResult } from '../../mcp/client';
import { getPythonScript, runPythonScript } from './python-runtime';
import { capToolOutput } from './output-cap';

export class SecurityTool implements BuiltInTool {
  name = 'security';
  description = '授权范围内的网络安全侦察与合规审计（Python 版，仅限自有资产或已授权目标）。action: port_scan(TCP 端口扫描+服务推断) / http_headers(HTTP 安全响应头审计 HSTS/CSP/X-Frame-Options…) / ssl_check(TLS 证书有效期与颁发者审计) / dns_lookup(DNS 记录 A/AAAA/CNAME/MX/NS/TXT) / subdomain(字典子域发现，仅 DNS 解析)。不含任何攻击性动作。';

  inputSchema = {
    type: 'object',
    properties: {
      action: {
        type: 'string',
        enum: ['port_scan', 'http_headers', 'ssl_check', 'dns_lookup', 'subdomain'],
        description: '侦察/审计动作。',
      },
      host: { type: 'string', description: '目标主机（port_scan/ssl_check 用，IP 或域名）。必须授权目标。' },
      ports: { type: 'string', description: 'port_scan 端口规格："common" / "80,443" / "1-1024"。' },
      timeout: { type: 'number', description: '超时（port_scan 毫秒 / 其他秒）。' },
      concurrency: { type: 'number', description: 'port_scan 并发数（默认 200）。' },
      url: { type: 'string', description: 'http_headers 的目标 URL。' },
      port: { type: 'number', description: 'ssl_check 的端口（默认 443）。' },
      name: { type: 'string', description: 'dns_lookup 的域名。' },
      type: { type: 'string', description: 'dns_lookup 记录类型（A/AAAA/CNAME/MX/NS/TXT，默认 A）。' },
      domain: { type: 'string', description: 'subdomain 的父域名。' },
    },
    required: ['action'],
  };

  async execute(args: Record<string, unknown>): Promise<McpCallResult> {
    const action = args.action as string;
    const script = getPythonScript('security/security.py');
    if (!script) {
      return { content: [{ type: 'text', text: 'Error: 未找到 security.py 脚本（resources/python-tools 或 packages/core 开发目录）。' }], isError: true };
    }

    const argv: string[] = [action];
    try {
      switch (action) {
        case 'port_scan': {
          const host = args.host as string;
          if (!host) throw new Error('port_scan 需要 host 参数');
          argv.push('--host', host);
          if (args.ports) argv.push('--ports', String(args.ports));
          if (args.timeout) argv.push('--timeout', String(args.timeout));
          if (args.concurrency) argv.push('--concurrency', String(args.concurrency));
          break;
        }
        case 'http_headers': {
          const url = args.url as string;
          if (!url) throw new Error('http_headers 需要 url 参数');
          argv.push('--url', url);
          if (args.timeout) argv.push('--timeout', String(args.timeout));
          break;
        }
        case 'ssl_check': {
          const host = args.host as string;
          if (!host) throw new Error('ssl_check 需要 host 参数');
          argv.push('--host', host);
          if (args.port) argv.push('--port', String(args.port));
          if (args.timeout) argv.push('--timeout', String(args.timeout));
          break;
        }
        case 'dns_lookup': {
          const name = args.name as string;
          if (!name) throw new Error('dns_lookup 需要 name 参数');
          argv.push('--name', name);
          if (args.type) argv.push('--type', String(args.type));
          break;
        }
        case 'subdomain': {
          const domain = args.domain as string;
          if (!domain) throw new Error('subdomain 需要 domain 参数');
          argv.push('--domain', domain);
          break;
        }
        default:
          throw new Error(`未知 action: ${action}`);
      }
    } catch (e) {
      return { content: [{ type: 'text', text: `Error: ${(e as Error).message}` }], isError: true };
    }

    try {
      const r = await runPythonScript(script, argv, { timeout: 120000 });
      const out = [r.stdout ? capToolOutput(r.stdout) : '', r.stderr ? '[stderr]\n' + capToolOutput(r.stderr) : '']
        .filter(Boolean)
        .join('\n') || '(无输出)';
      if (r.exitCode !== 0) {
        return { content: [{ type: 'text', text: `security ${action} 失败（exit ${r.exitCode}）：\n${out}` }], isError: true };
      }
      return { content: [{ type: 'text', text: out }] };
    } catch (e: unknown) {
      const msg = e instanceof Error ? e.message : String(e);
      return { content: [{ type: 'text', text: `security 执行异常: ${msg}` }], isError: true };
    }
  }
}
